### PR TITLE
convert more functions in directoryDao to IO

### PR DESCRIPTION
**Pull request by Qi77Qi**
Ticket: No ticket
Main change is in https://github.com/broadinstitute/sam/compare/ldap?expand=1#diff-010567be56b84d5efdd2665fa097c7ce

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 1]
```diff
@@ -95,16 +93,16 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
...
+    cs.evalOn(ecForLdapBlockingIO)(IO(
       ldapConnectionPool.modify(groupDn(groupId),
         new Modification(ModificationType.DELETE, Attr.uniqueMember, subjectDn(removeMember)),
         groupUpdatedModification
       )
-    } match {
-      case Success(_) => true
-      case Failure(ldape: LDAPException) if ldape.getResultCode == ResultCode.NO_SUCH_ATTRIBUTE => false
-      case Failure(regrets) => throw regrets
+    )).attempt.flatMap{
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: nit: space before curly brace

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 2]
```diff
@@ -153,46 +151,47 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
...
     }
   }
 
-  private def unmarshalUser(results: Entry): WorkbenchUser = {
-    val uid = getAttribute(results, Attr.uid).getOrElse(throw new WorkbenchException(s"${Attr.uid} attribute missing"))
-    val email = getAttribute(results, Attr.email).getOrElse(throw new WorkbenchException(s"${Attr.email} attribute missing"))
-
-    WorkbenchUser(WorkbenchUserId(uid), getAttribute(results, Attr.googleSubjectId).map(GoogleSubjectId), WorkbenchEmail(email))
+  private def unmarshalUser(results: Entry): Either[String, WorkbenchUser] = {
+    for{
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: curly brace space

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 3]
```diff
@@ -153,46 +151,47 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
...
       new Attribute(Attr.cn, user.id.value),
       new Attribute(Attr.uid, user.id.value),
       new Attribute("objectclass", Seq("top", "workbenchPerson").asJava)
     ) ++ user.googleSubjectId.map(gsid => List(new Attribute(Attr.googleSubjectId, gsid.value))).getOrElse(List.empty)
-    Future {
-      ldapConnectionPool.add(userDn(user.id), attrs: _*)
-      user
-    }.recoverWith {
+
+    cs.evalOn(ecForLdapBlockingIO)(IO(ldapConnectionPool.add(userDn(user.id), attrs: _*))).adaptError{
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: curly brace space

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 4]
```diff
@@ -153,46 +151,47 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
...
+      uid <- getAttribute(results, Attr.uid).toRight(s"${Attr.uid} attribute missing")
+      email <- getAttribute(results, Attr.email).toRight(s"${Attr.email} attribute missing")
+    } yield WorkbenchUser(WorkbenchUserId(uid), getAttribute(results, Attr.googleSubjectId).map(GoogleSubjectId), WorkbenchEmail(email))
   }
 
-  override def loadUsers(userIds: Set[WorkbenchUserId]): Future[Seq[WorkbenchUser]] = Future {
+  override def loadUsers(userIds: Set[WorkbenchUserId]): IO[Stream[WorkbenchUser]] = {
     val filters = userIds.grouped(batchSize).map(batch => Filter.createORFilter(batch.map(g => Filter.createEqualityFilter(Attr.uid, g.value)).asJava)).toSeq
-    ldapSearchStream(peopleOu, SearchScope.ONE, filters:_*)(unmarshalUser)
+    for{
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: curly brace space

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala [hunk 5]
```diff
@@ -48,18 +48,18 @@ class UserService(val directoryDAO: DirectoryDAO, val cloudExtensions: CloudExte
     *      yes            skip    ---> User exists. Do nothing.
     *      yes            skip    ---> User exists. Do nothing.
     */
-  protected[service] def registerUser(user: CreateWorkbenchUser): Future[WorkbenchUser] = for{
-    existingSubFromGoogleSubjectId <- directoryDAO.loadSubjectFromGoogleSubjectId(user.googleSubjectId).unsafeToFuture()
+  protected[service] def registerUser(user: CreateWorkbenchUser): IO[WorkbenchUser] = for{
```
Comment (marctalbott) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala]: curly brace space

**Review by marctalbott — APPROVED** [review]
Just a handful of places where it'd be nice to see a space before the curly brace

**Review by dvoet — APPROVED** [review]

**Comment by Qi77Qi** [comment]
jenkins retest

**Pull request merged**